src/main/Extensions/CustomWebSearch/CustomWebSearchExtension.ts

```diff
@@ -0,0 +1,135 @@
+import type { AssetPathResolver } from "@Core/AssetPathResolver";
+import type { Extension } from "@Core/Extension";
+import type { SettingsManager } from "@Core/SettingsManager";
+import type { OperatingSystem, SearchResultItem } from "@common/Core";
+import type { Image } from "@common/Core/Image";
+import type { CustomSearchEngineSetting, Settings } from "@common/Extensions/CustomWebSearch";
+
+export class CustomWebSearchExtension implements Extension {
+    public readonly id = "CustomWebSearch";
+    public readonly name = "Browser Search";
+
+    public readonly nameTranslation = {
+        key: "extensionName",
+        namespace: "extension[CustomWebSearch]",
+    };
+
+    public readonly author = {
+        name: "NiewView",
+        githubUserName: "NiewView",
+    };
+
+    public constructor(
+        private readonly operatingSystem: OperatingSystem,
+        private readonly assetPathResolver: AssetPathResolver,
+        private readonly settingsManager: SettingsManager,
+    ) {}
+
+    async getSearchResultItems(): Promise<SearchResultItem[]> {
+        // Custom search engines do not have static search results
+        return [];
+    }
+
+    public getInstantSearchResultItems(searchTerm: string): SearchResultItem[] {
+        const customSearchEngines = this.settingsManager.getValue<CustomSearchEngineSetting[]>(
+            `extension[${this.id}].customSearchEngines`,
+            this.getSettingDefaultValue("customSearchEngines"),
+        );
+
+        const selectedSearchEngine = customSearchEngines.find((engine) => searchTerm.startsWith(engine.prefix));
+
+        if (!selectedSearchEngine) {
+            return [];
+        }
+
+        const searchInput = searchTerm.replace(selectedSearchEngine.prefix, "").trim();
+        const encodeSearchInput = selectedSearchEngine.encodeSearchTerm ? encodeURIComponent(searchInput) : searchInput;
+
+        return [
+            {
+                name: selectedSearchEngine.name,
+                description: `Search in ${selectedSearchEngine.name}`,
+                id: `${selectedSearchEngine.name}:instantResult`,
+                image: this.getImage(),
```

Comment:
Instead of this static image we could use the favicon from the URL directly. Then the search results would look like this, what do you think?

<img width="549" alt="image" src="https://github.com/user-attachments/assets/8683fe7d-8e8a-4d38-b527-30b15cd7dcb2">

You could do this by adding the `UrlImageGenerator` to the constructor:

```typescript
public constructor(
    private readonly operatingSystem: OperatingSystem,
    private readonly assetPathResolver: AssetPathResolver,
    private readonly settingsManager: SettingsManager,
    private readonly urlImageGenerator: UrlImageGenerator,
) {}
```

Then use it when creating the search result items:

```typescript
return [
    {
        name: selectedSearchEngine.name,
        description: `Search in ${selectedSearchEngine.name}`,
        id: `${selectedSearchEngine.name}:instantResult`,
        image: this.urlImageGenerator.getImage(new URL(selectedSearchEngine.url).origin),
        defaultAction: {
            handlerId: "Url",
            description: `Search in default Browser`,
            argument: selectedSearchEngine.url.replace("{{query}}", encodeSearchInput),
        },
    },
];
```

Reply:
That's great, I thought it will need way more effort to get it. 
I used the resolved URL for image fetching, so it will also work, if the query is part of the domain name.